content/part2.md

```diff
@@ -0,0 +1,186 @@
+
+Sending Emails from an Activity
+-------------------------------
+
+You can think of Activities as an abstraction for side effects in Temporal.
+[Workflows should be pure, idempotent functions](https://docs.temporal.io/docs/go-create-workflows/#implementation) to allow Temporal to re-run a Workflow to recreate the Workflow's state.
+Any side effects, like HTTP requests to the [Mailgun API](https://thecodebarbarian.com/sending-emails-using-the-mailgun-api.html), should be in an Activity.
+
+For example, below is the implementation of the `SendAbandonedCartEmail()` function.
+It loads Mailgun keys from environment variables, and sends an HTTP request to the Mailgun API using [Mailgun's official Go library](https://github.com/mailgun/mailgun-go).
+The function takes two parameters: the workflow context, and the email as a string.
+
+```golang
+import (
+	"context"
+	"fmt"
+	"github.com/mailgun/mailgun-go"
+)
+
+var (
+	mailgunDomain = os.Getenv("MAILGUN_DOMAIN")
+	mailgunKey    = os.Getenv("MAILGUN_PRIVATE_KEY")
+)
+
+func SendAbandonedCartEmail(_ context.Context, email string) error {
+	mg := mailgun.NewMailgun(mailgunDomain, mailgunKey)
+	m := mg.NewMessage(
+		"noreply@"+mailgunDomain, // Sender
+		"You've abandoned your shopping cart!", // Subject
+		"Go to http://localhost:8080 to finish checking out!", // Placeholder email copy
+		email, // Recipient
+	)
+	_, _, err := mg.Send(m)
+	if err != nil {
+		fmt.Println("Mailgun err: " + err.Error())
+    return err
+	}
+
+	return err
+}
+```
+
+As a reminder, below is the `ExecuteActivity()` call from the cart Workflow.
+The 3rd parameter to `ExecuteActivity()` becomes the 2nd parameter to `SendAbandonedCartEmail()`:
+
+```golang
+workflow.ExecuteActivity(ctx, SendAbandonedCartEmail, state.Email).Get(ctx, nil)
+```
+
+The `ExecuteActivity()` function also exposes some neat options.
+For example, [Temporal automatically retries failed activities](https://docs.temporal.io/docs/go-retries/), so Temporal would automatically retry the `SendAbandonedCart()` Activity up to 5 times if `SendAbandonedCart()` returns an error.
+You can configure the number of times Temporal will retry the Activity in case of an error using the `MaximumAttempts` option as shown below.
+
+```golang
+ao := workflow.ActivityOptions{
+	ScheduleToStartTimeout: time.Minute,
+	StartToCloseTimeout:    time.Minute,
+  MaximumAttempts:        3,
+}
+
+ctx = workflow.WithActivityOptions(ctx, ao)
+
+err := workflow.ExecuteActivity(ctx, SendAbandonedCartEmail, state.Email).Get(ctx, nil)
+```
+
+Moving On
+---------
+
+Long-lived Workflows in Temporal are excellent for scheduled tasks.
+You can build durable time-based logic, like checking whether the user hasn't modified their shopping cart for a given period of time, without using a job queue.
+Next up, we'll look at patterns for build RESTful APIs on top of Temporal Workflows.
```

Comment:
```suggestion
Next up, we'll look at patterns for building RESTful APIs on top of Temporal Workflows.
```

Reply:
Applied in https://github.com/vkarpov15/temporal-ecommerce/pull/3/commits/ee942b604506e05eca45580409006b6057e48fa8, can't merge through GitHub UI because there's a merge conflict somewhere.